modules/infra/control/mempool.c

```diff
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include "gr_mempool.h"
+
+#include <gr_log.h>
+#include <gr_mbuf.h>
+
+struct mempool_tracker {
+	struct rte_mempool *mp;
+	uint32_t reserved;
+};
+
+#define MAX_MEMPOOL_PER_NUMA 32
+#define MEMPOOL_DEFAULT_SIZE (1 << 16) - 1
+
+static int mt_sort(const void *p1, const void *p2) {
+	const struct mempool_tracker *mt1 = p1;
+	const struct mempool_tracker *mt2 = p2;
+
+	if (mt1->mp == mt2->mp)
+		return 0;
+
+	if (mt1->mp && mt2->mp) {
+		return (mt2->mp->size - mt2->reserved) - (mt1->mp->size - mt1->reserved);
+	} else {
+		return (mt1->mp ? -1 : 1);
+	}
+}
+
+static struct mempool_tracker mempools[RTE_MAX_NUMA_NODES + 1][MAX_MEMPOOL_PER_NUMA] = {0};
+static uint32_t mempool_default_size = MEMPOOL_DEFAULT_SIZE;
+
+struct rte_mempool *gr_pktmbuf_pool_get(int8_t socket_id, uint32_t count) {
+	char mp_name[RTE_MEMPOOL_NAMESIZE];
+	struct rte_mempool *mp = NULL;
+	uint32_t alloc_size;
+
+	for (int i = 0; i < MAX_MEMPOOL_PER_NUMA; i++) {
+		struct mempool_tracker *mt = &mempools[socket_id + 1][i];
+		if (mt->mp == NULL) {
+			alloc_size = mempool_default_size;
+			if (count > mempool_default_size / 4) {
+				alloc_size = count * 2;
+				alloc_size = rte_align32pow2(alloc_size) - 1;
+				// For future mempools, increase default size;
+				mempool_default_size = alloc_size;
+			}
+			sprintf(mp_name, "mp_gr_%d:%d", socket_id, i);
+			LOG(DEBUG, "Allocating new mempool '%s' with size 0x%x", mp_name, alloc_size
+			);
```

Comment:
`clang-format` chokes when the line size is close to the limit with short function names...

Maybe change size format to %u and/or shorten the message:

```c
                 LOG(DEBUG, "alloc mempool %s size %u", mp_name, alloc_size);
```